homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,113 @@
+"""Module to coordinate llm tools."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Iterable
+from dataclasses import dataclass
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.weather.intent import INTENT_GET_WEATHER
+from homeassistant.core import Context, HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.util.json import JsonObjectType
+
+from . import intent
+
+_LOGGER = logging.getLogger(__name__)
+
+IGNORE_INTENTS = [intent.INTENT_NEVERMIND, intent.INTENT_GET_STATE, INTENT_GET_WEATHER]
+
+
+@dataclass(slots=True)
+class ToolInput:
+    """Tool input to be processed."""
+
+    tool_name: str
+    tool_args: dict[str, Any]
+    platform: str
+    context: Context | None
+    user_prompt: str | None
+    language: str | None
+    assistant: str | None
+
+
+class Tool:
+    """LLM Tool base class."""
+
+    name: str
+    description: str | None = None
+    parameters: vol.Schema = vol.Schema({})
+
+    @abstractmethod
+    async def async_call(
+        self, hass: HomeAssistant, tool_input: ToolInput
+    ) -> JsonObjectType:
+        """Call the tool."""
+        raise NotImplementedError
+
+    def __repr__(self) -> str:
+        """Represent a string of a Tool."""
+        return f"<{self.__class__.__name__} - {self.name}>"
+
+
+@callback
+def async_get_tools(hass: HomeAssistant) -> Iterable[Tool]:
+    """Return a list of LLM tools."""
+    for intent_handler in intent.async_get(hass):
+        if intent_handler.intent_type not in IGNORE_INTENTS:
+            yield IntentTool(intent_handler)
+
+
+@callback
+async def async_call_tool(hass: HomeAssistant, tool_input: ToolInput) -> JsonObjectType:
+    """Call a LLM tool, validate args and return the response."""
+    for tool in async_get_tools(hass):
+        if tool.name == tool_input.tool_name:
+            break
+    else:
+        raise HomeAssistantError(f'Tool "{tool_input.tool_name}" not found')
+
+    if tool_input.context is None:
+        tool_input.context = Context()
+
+    tool_input.tool_args = tool.parameters(tool_input.tool_args)
+
+    return await tool.async_call(hass, tool_input)
+
+
+class IntentTool(Tool):
+    """LLM Tool representing an Intent."""
+
+    def __init__(
+        self,
+        intent_handler: intent.IntentHandler,
+    ) -> None:
+        """Init the class."""
+        if intent_handler.intent_type is None:
+            raise HomeAssistantError("intent type cannot be None")
+
+        self.name = intent_handler.intent_type
+        self.description = f"Execute Home Assistant {self.name} intent"
```

Comment:
Idea for future: Names are currently self explanatory but it may be useful in the future to add the description from the docs at https://developers.home-assistant.io/docs/intent_builtin/ e.g. "Adds an item to the shopping list" vs "Execute Home Assistant HassShoppingListAddItem intent"

Reply:
Yes, good point. I am already thinking about adding an optional attribute to the intent handler.